engine/kv_engine_cleaner.cpp

```diff
@@ -235,366 +237,431 @@ void KVEngine::cleanNoHashIndexedSkiplist(
+  }
 
-      {  // Destroy hash
-        while (!outdated_hash_lists.empty()) {
-          auto& ts_hlist = outdated_hash_lists.front();
-          if (ts_hlist.first < version_controller_.LocalOldestSnapshotTS()) {
-            hashListDestroy(ts_hlist.second.release());
-            outdated_hash_lists.pop_front();
-          } else {
-            break;
-          }
-        }
-      }
-    }  // Finsh iterating hash table
+  if (!purge_dl_records.empty()) {
+    pending_clean_records.pending_purge_dls.emplace_back(
+        std::move(purge_dl_records), new_ts);
+    purge_dl_records = std::vector<DLRecord*>();
+  }
 
-    // Push the remaining need purged records to global pool.
-    auto new_ts = version_controller_.GetCurrentTimestamp();
-    if (!purge_string_records.empty()) {
-      pending_purge_strings.emplace_back(std::move(purge_string_records),
-                                         new_ts);
-      purge_string_records = std::vector<StringRecord*>();
+  return total_num == 0 ? 0.0f : need_purge_num / (double)total_num;
+}
+
+void KVEngine::TestCleanOutDated(size_t start_slot_idx, size_t end_slot_idx) {
+  PendingCleanRecords pending_clean_records;
+  while (!bg_work_signals_.terminating) {
+    auto cur_slot_idx = start_slot_idx;
+    while (cur_slot_idx < end_slot_idx && !bg_work_signals_.terminating) {
+      cleanSlotBlockOutDated(pending_clean_records, cur_slot_idx, 1024);
+      cur_slot_idx += 1024;
     }
+  }
+}
 
-    if (!purge_dl_records.empty()) {
-      pending_purge_dls.emplace_back(std::move(purge_dl_records), new_ts);
-      purge_dl_records = std::vector<DLRecord*>();
+// Space Reclaimer
+void SpaceReclaimer::addNewWorker(size_t thread_id) {
+  PendingCleanRecords pending_clean_records;
+  while (true) {
+    if (close_) return;
+    cur_slot_idx_ = (cur_slot_idx_ + kSlotBlockUnit) %
+                    kv_engine_->hash_table_->GetSlotsNum();
+    kv_engine_->cleanSlotBlockOutDated(pending_clean_records,
+                                       cur_slot_idx_.load(), kSlotBlockUnit);
+
+    bool finish = false;
+    {
+      std::unique_lock<std::mutex> worker_lock(workers_[thread_id].mtx);
+      finish = workers_[thread_id].finish;
+    }
+    if (finish) {
+      while (pending_clean_records.Size() != 0 && !close_.load()) {
+        kv_engine_->version_controller_.UpdateLocalOldestSnapshot();
+        kv_engine_->prugeAndFreeAllType(pending_clean_records);
+      }
+      return;
+    }
+  }
+}
+
+void SpaceReclaimer::AdjustThread(size_t advice_thread_num) {
+  auto active_thread_num = live_thread_num_.load();
+  if (active_thread_num < advice_thread_num) {
+    for (size_t i = active_thread_num; i < advice_thread_num; ++i) {
+      size_t thread_id = idled_workers_.front();
+      idled_workers_.pop_front();
+      std::thread worker(&SpaceReclaimer::addNewWorker, this, thread_id);
+      workers_[thread_id].finish = false;
+      workers_[thread_id].worker = std::move(worker);
```

Comment:
no need to move, workers_[thread_id].worker = std::thread(&SpaceReclaimer::addNewWorker, this, thread_id);